File: src/test/java/com/github/comius/floats/FloatsTest.java
package com.github.comius.floats;

import static org.junit.Assert.assertTrue;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertEquals;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.CoreMatchers.nullValue;
import static org.hamcrest.CoreMatchers.notNullValue;
import java.math.MathContext;
import java.math.RoundingMode;
import java.util.function.BiFunction;
import java.util.function.IntPredicate;
import java.util.Optional;

import org.junit.Test;
import org.junit.Rule;
import org.junit.rules.ErrorCollector;

import com.github.comius.floats.Floats;
import com.github.comius.floats.Floats.Float;

/**
 * Unit tests for Floats.<br>
 * 
 * Floats are tested trough the Floats interface. Currently BigDecimal implementation is tested, but also any other
 * implementation could be tested similarly.
 * 
 * <p>
 * Environment: no specific test environment needs to be set up. Java/Scala provide everything in default installation.
 *
 */
public class FloatsTest {
    @Rule
    public ErrorCollector collector = new ErrorCollector();

    /** An integer repesenting the infinity. */
    private int inf = 100;

    // Next two arrays represent first and second operand for int operation and third array is the same operand for
    // Float operation. Operation can be addition, subtraction, multiplication or division.
    private int[] testValuesInt1 = new int[] { inf + 5, -inf - 5, 0, 1, -1 };
    private int[] testValuesInt2 = new int[] { inf + 10, -inf - 10, 0, 1, -1 };
    private Float[] testValuesDyadic = new Float[] { Floats.impl().posInf(), Floats.impl().negInf(),
	    Floats.impl().ZERO(), Floats.impl().valueOf(1), Floats.impl().valueOf(-1) };

    /** Normalizes infinities. Anything bigger or smaller is truncated to infinity. */
    private int normalize(int a) {
	if (a >= inf)
	    return inf;
	if (a <= -inf)
	    return -inf;
	return a;
    }

    /**
     * Tests limits of given binary arithmetic operation.
     * 
     * @param op
     *            the arithmetic operation on Floats
     * @param intOp
     *            the same arithmetic operation on integers
     * @param opStr
     *            the string describing the operation
     */
    private void testLimits(BiFunction<Float, Float, Float> op, BiFunction<Integer, Integer, Integer> intOp,
	    String opStr) {
	for (int i = 0; i < testValuesInt1.length; i++) {
	    for (int j = 0; j < testValuesInt1.length; j++) {
		// Computes value using integers and embeds it into Floats
		Optional<Float> cprime;
		try {
		    int ai = testValuesInt1[i];
		    int bi = testValuesInt2[j];
		    int ci = normalize(intOp.apply(ai, bi));
		    if (ci >= inf)
			cprime = Optional.of(Floats.impl().posInf());
		    else if (ci <= -inf)
			cprime = Optional.of(Floats.impl().negInf());
		    else if (ci > 2 || ci < -2)
			cprime = Optional.empty();
		    else
			cprime = Optional.of(Floats.impl().valueOf(ci));
		} catch (ArithmeticException e) {
		    cprime = Optional.empty();
		}

		// The operation on floats
		Float a = testValuesDyadic[i];
		Float b = testValuesDyadic[j];
		try {
		    Float c = op.apply(a, b);
		    String logMessage = a + opStr + b + " = " + c + ": expected NaN";

		    // Verifies if the result matches
		    collector.checkThat(logMessage, cprime.orElse(null), notNullValue());
		    collector.checkThat(logMessage, c, equalTo(cprime.get()));
		} catch (ArithmeticException e) {
		    String logMessage = a + opStr + b + " = NaN: " + cprime;

		    // Verifies both results are NaN
		    collector.checkThat(logMessage + e.getMessage(), cprime.orElse(null), nullValue());
		}
	    }
	}

    }

    /**
     * Tests limits of arithmetic operations.
     */
    @Test
    public void testLimitsOfArithmeticOperations() {
	// Test addition
	testLimits((a, b) -> a.add(b, new MathContext(10, RoundingMode.CEILING)), (a, b) -> a + b, " + ");

	// Test subtraction
	testLimits((a, b) -> a.subtract(b, new MathContext(10, RoundingMode.CEILING)), (a, b) -> a - b, " - ");

	IntPredicate isInf = a -> Math.abs(normalize(a)) == inf;

	// Multiplication
	// Special case multiplication inf and 0
	BiFunction<Integer, Integer, Integer> multiply = (a,
		b) -> (isInf.test(a) && b == 0) || (isInf.test(b) && a == 0) ? 1 / 0 : a * b;
	testLimits((a, b) -> a.multiply(b, new MathContext(10, RoundingMode.CEILING)), multiply, " * ");

	// Division
	// Special case division of infinities
	BiFunction<Integer, Integer, Integer> divide = (a, b) -> isInf.test(a) && isInf.test(b) ? 1 / 0 : a / b;
	testLimits((a, b) -> a.divide(b, new MathContext(10, RoundingMode.CEILING)), divide, " / ");
    }

    @Test
    public void testExponent() {
	MathContext mc = new MathContext(1, RoundingMode.FLOOR);
	Float a = Floats.impl().valueOf("49e2147483647", mc);
	System.out.println(a.add(a, mc));
	System.out.println(Floats.impl().valueOf("17", new MathContext(2, RoundingMode.FLOOR)));
    }
}

File: src/test/scala/com/github/comius/floats/FloatsTest.scala
/*
 * Dedekind Reals - Java Library for computing with Dedekind Reals
 * Copyright (c) 2019 Ivo List
 *
 * This software is distributed under the terms found
 * in file LICENSE.txt that is included with this distribution.
 */

package com.github.comius.floats

import java.math.MathContext
import java.math.RoundingMode

import scala.reflect.ClassTag

import org.junit.Test

import com.github.comius.floats.BigDecimalFloats.BigDecimalFloat
import com.github.comius.floats.Floats.{ impl => D }

/**
 * Unit tests for Floats.
 *
 * Floats are tested trough the Floats interface. Currently BigDecimal implementation is tested, but also any other
 * implementation could be tested similarly.
 *
 * Environment: no specific test environment needs to be set up. Java/Scala provide everything in default installation.
 *
 */
// scalastyle:off magic.number
// scalastyle:off regex
class FloatsTest {

  /** An integer representing the infinity. */
  val inf = 100

  // Next two arrays represent first and second operand for int operation and third array is the same operand for
  // Float operation. Operation can be addition, subtraction, multiplication or division.
  val testValuesInt1 = Array(inf + 5, -inf - 5, 0, 1, -1)
  val testValuesInt2 = Array(inf + 10, -inf - 10, 0, 1, -1)

  implicit val implClass: ClassTag[D.T] = ClassTag(classOf[BigDecimalFloat])
  val testValuesDyadic = Array(D.posInf, D.negInf, D.ZERO, D.valueOf(1), D.valueOf(-1))

  /** Normalizes infinities. Anything bigger or smaller is truncated to infinity. */
  def normalize(a: Int): Int = {
    if (a >= inf) {
      inf
    } else if (a <= -inf) {
      -inf
    } else {
      a
    }
  }

  /**
   * Tests limits of given binary arithmetic operation.
   *
   * @param op
   *            the arithmetic operation on Floats
   * @param intOp
   *            the same arithmetic operation on integers
   * @param opStr
   *            the string describing the operation
   */
  def testLimits(op: (D.T, D.T) => D.T, intOp: (Int, Int) => Int, opStr: String): Unit = {
    for { i <- 0 until testValuesInt1.length } {
      for { j <- 0 until testValuesInt1.length } {
        // Computes value using integers and embeds it into Floats
        val cprime: Option[D.T] =
          try {
            val ai = testValuesInt1(i)
            val bi = testValuesInt2(j)
            val ci: Int = normalize(intOp(ai, bi))
            val maxResult = 2

            if (ci >= inf) {
              Some(D.posInf)
            } else if (ci <= -inf) {
              Some(D.negInf)
            } else if (ci > maxResult || ci < -maxResult) {
              None
            } else {
              Some(D.valueOf(ci))
            }
          } catch {
            case e: ArithmeticException =>
              None
          }

        // The operation on floats
        val a = testValuesDyadic(i);
        val b = testValuesDyadic(j);
        try {
          val c: D.T = op(a, b);
          val logMessage = a + opStr + b + " = " + c + ": expected NaN";

          // Verifies if the result matches
          assert(cprime.isDefined, logMessage);
          assert(c == cprime.get, logMessage);
        } catch {
          case e: ArithmeticException =>
            val logMessage = a + opStr + b + " = NaN: " + cprime;

            // Verifies both results are NaN
            assert(cprime.isEmpty, logMessage);
        }
      }
    }

  }

  /**
   * Tests limits of arithmetic operations.
   */
  @Test
  def testLimits(): Unit = {
    val precision = 10

    // Test addition
    testLimits((a, b) => a.add(b, new MathContext(precision, RoundingMode.CEILING)), (a, b) => a + b, " + ");

    // Test subtraction
    testLimits((a, b) => a.subtract(b, new MathContext(precision, RoundingMode.CEILING)), (a, b) => a - b, " - ");

    val isInf = (a: Int) => Math.abs(normalize(a)) == inf;

    // Multiplication
    // Special case multiplication inf and 0
    val multiply = (a: Int, b: Int) =>
      if ((isInf(a) && b == 0) || (isInf(b) && a == 0)) {
        throw new ArithmeticException()
      } else { a * b }
    testLimits((a, b) => a.multiply(b, new MathContext(precision, RoundingMode.CEILING)), multiply, " * ");

    // Division
    // Special case division of infinities
    val divide = (a: Int, b: Int) => if (isInf(a) && isInf(b)) throw new ArithmeticException() else a / b;
    testLimits((a, b) => a.divide(b, new MathContext(precision, RoundingMode.CEILING)), divide, " / ");
  }

  def getPrec(a: Int): Int = {
    var p = 10
    if (a == 0) return p;
    var aa = a
    while ((aa % 2) == 0) {
      aa = aa / 2
      p = p - 1;
    }
    p
  }

  /**
   * Empirical tests of trisection.
   *
   * First test splits at 1/4 and 1/4 - assumption is that we can keep precision lower, to be more efficient.
   *
   * Second test splits at 1/2-eps and 1/2+eps - assumption is that we are more efficient by having less iterations.
   *
   * Third test splits at 2 numbers with lowest precision, at least it tries.
   *
   * Conclusion: Second test always wins by having less iteration, because other tests have tail iterations with highest
   * precision, which don't balance with initial iterations at lower precision.
   */
  def testTrisection(): Unit = {
    val s = Integer.parseInt("00001010100", 2);
    var a = 0
    var b = Integer.parseInt("100000000000", 2);
    var i0 = 0;
    var psum = 0;
    while (b - a > 2) {
      var a1 = (3 * a + b) / 4
      var b1 = (a + 3 * b) / 4

      if (a1 < s) a = a1;
      if (s < b1) b = b1;
      println(s" iteration $i0 precision $a1 $b1 ${getPrec(a1)} ${getPrec(b1)}")
      psum += getPrec(a1) + getPrec(b1);
      i0 += 1
    }
    println(s"$a $b $s")
    println(psum);
  }

  def testTrisection2(): Unit = {
    val s = Integer.parseInt("00001010100", 2);
    var a = 0
    var b = Integer.parseInt("100000000000", 2);
    var i = 0;
    var psum = 0;
    while (b - a > 2) {
      val a1 = (a + b) / 2 - 1
      val b1 = (a + b) / 2 + 1
      if (a1 < s) a = a1;
      if (s < b1) b = b1;
      println(s" iteration $i precision $a1 $b1  ${getPrec(a1)} ${getPrec(b1)}")
      psum += getPrec(a1) + getPrec(b1);
      i += 1
    }
    println(s"$a $b $s")
    println(psum);
  }

  def testTrisection3(): Unit = {
    val s = Integer.parseInt("00001010100", 2);
    var a = 0
    var b = Integer.parseInt("100000000000", 2);
    var i0 = 0;
    var psum = 0;
    var p = b
    while (b - a > 2) {
      var a1 = (a + b) / 2
      var b1 = (a + b) / 2
      while (a1 - p <= a && b1 + p >= b) p /= 2
      a1 = a1 - p
      b1 = b1 + p

      if (a1 < s) a = a1;
      if (s < b1) b = b1;
      println(s" iteration $i0 precision $a1 $b1 ${getPrec(a1)} ${getPrec(b1)} $p")
      psum += getPrec(a1) + getPrec(b1);
      i0 += 1
    }
    println(s"$a $b $s")
    println(psum);
  }
}

